### PR TITLE
Added support for Multipart form-data in OpenAPI addon 

### PR DESCRIPTION
**Pull request by pranavsaxena17**
Fixes zaproxy/zaproxy#6418

### CODE REVIEW

**Comment by ricekot** [comment]
Thanks for tackling this! Can you rebase these changes on top of the latest commits?

**Comment by pranavsaxena17** [comment]
@ricekot Done

**Comment by thc202** [comment]
Any reason for the use of a real image (assuming it is) rather than dummy data?

**Comment by pranavsaxena17** [comment]
Nah, it's just the ZAP icon so I put it in as kind of an easter egg. Should I use dummy data instead?

**Comment by thc202** [comment]
That's preferable, this not only affects the code but the HTTP requests done later when importing/scanning, so better if we use less data by default. IMO we could just add the field without data. If the user needs are real/specific image they should send the request with that data, instead of relying on default values.

The default text could also be reduced to a single line (or even word).

**Comment by pranavsaxena17** [comment]
@thc202 Done

File: addOns/openapi/CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Support for statistics (number of URLs added) 
 
 ### Changed
+- Added support for Multipart form-data (Issue 6418).
```
Comment (thc202) [inline addOns/openapi/CHANGELOG.md:17]: This should be added to the unreleased section.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 2]
```diff
@@ -77,8 +82,10 @@ public String generate(Schema<?> schema) {
         boolean isArray = schema instanceof ArraySchema;
         LOG.debug("Generate body for object {}", schema.getName());
 
-        if (isArray) {
+        if (schema instanceof ArraySchema) {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:85]: Why this change, or why was not `isArray` removed?

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java [hunk 3]
```diff
@@ -60,7 +60,16 @@ public Requestor(int initiator) {
...
+                    if (hhf.getValue().contains("multipart/form-data")) {
+                        String requestBody = requestModel.getBody();
+                        String boundary = requestBody.substring(2, 38);
+                        httpRequest
+                                .getRequestHeader()
+                                .setHeader(
+                                        hhf.getName(), hhf.getValue() + "; boundary=" + boundary);
+                    } else {
+                        httpRequest.getRequestHeader().setHeader(hhf.getName(), hhf.getValue());
+                    }
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java:72]: This should be reverted, it's not responsibility of the `Requestor` to tweak/correct generated data.

File: addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/generators/HeadersGeneratorUnitTest.java [hunk 4]
```diff
@@ -87,7 +87,7 @@ void shouldNotGenerateContentTypeIfRequestHaveNoMediaTypes() {
 
     @ParameterizedTest
     @ValueSource(strings = {"text/plain", "application/xml"})
-    void shouldNotGenerateContentTypeIfNotJsonNorWwwFormUrlEncoded(String mediaType) {
+    void shouldNotGenerateContentTypeIfNotJsonOrWwwFormUrlEncodedOrMultipart(String mediaType) {
```
Comment (thc202) [inline addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/generators/HeadersGeneratorUnitTest.java:90]: At this point might as well make it generic, e.g. `shouldNotGenerateContentTypeIfNotSupported`.

File: addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/generators/HeadersGeneratorUnitTest.java [hunk 5]
```diff
@@ -125,11 +125,27 @@ void shouldGenerateWwwFormUrlEncodedContentType(String mediaType) {
...
+
     @Test
     void shouldGenerateJustFirstSupportedContentType() {
         // Given
         RequestBody request =
-                mockRequestWithMediaTypes("application/json", "application/x-www-form-urlencoded");
+                mockRequestWithMediaTypes(
+                        "application/json",
+                        "application/x-www-form-urlencoded",
+                        "multipart/form-data");
```
Comment (thc202) [inline addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/generators/HeadersGeneratorUnitTest.java:148]: Should be reverted, two is enough to verify that the first is used.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 6]
```diff
@@ -122,6 +129,22 @@ private String generateFromArraySchema(ArraySchema schema) {
...
     }
 
+    private String generateFromBinarySchema(BinarySchema schema, Boolean isImage) {
+        if (isImage) {
+
+            // Dummy data
+            byte[] data = new byte[] {110, -12, 34, -18, 11, 69, 20, 11, 51, 26, 27, 14};
+
+            String imageFileContents = new String(data);
+            return imageFileContents;
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:139]: - The value can be returned without the variable.
- Should use a `Charset` to create the string.
- Should be a constant, not necessary to create a new string each time (it can also be public to be used in the tests).

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 7]
```diff
@@ -122,6 +129,22 @@ private String generateFromArraySchema(ArraySchema schema) {
         return createJsonArrayWith(generate(schema.getItems()));
     }
 
+    private String generateFromBinarySchema(BinarySchema schema, Boolean isImage) {
+        if (isImage) {
+
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:134]: These empty lines could be removed.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 8]
```diff
@@ -122,6 +129,22 @@ private String generateFromArraySchema(ArraySchema schema) {
         return createJsonArrayWith(generate(schema.getItems()));
     }
 
+    private String generateFromBinarySchema(BinarySchema schema, Boolean isImage) {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:132]: - `static`
 - `Boolean` → `boolean`

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 9]
```diff
@@ -122,6 +129,22 @@ private String generateFromArraySchema(ArraySchema schema) {
...
+        if (isImage) {
+
+            // Dummy data
+            byte[] data = new byte[] {110, -12, 34, -18, 11, 69, 20, 11, 51, 26, 27, 14};
+
+            String imageFileContents = new String(data);
+            return imageFileContents;
+
+        } else {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:141]: Remove the else statement, not necessary.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 10]
```diff
@@ -122,6 +129,22 @@ private String generateFromArraySchema(ArraySchema schema) {
...
+            String textFileContents =
+                    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus eu tortor efficitur";
+            return textFileContents;
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:144]: Similar comments as for the above value.

**Comment by thc202** [comment]
IMO the changes done to existing definitions should be reverted, or just remove the endpoint (since they are already being tested elsewhere).

**Comment by thc202** [comment]
I didn't check the actual generation of the data yet.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java (hunk 2 above)
Comment (thc202) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:85]: I see it was reintroduced when merging changes, so just needs to be removed again.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java (hunk 8 above)
Comment (pranavsaxena17) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:132]: @thc202 Making this method static would require making the public strings `imageFileContents` and `textFileContents` static too. Then they cannot be referenced in the tests (as `generators.getBodyGenerator()` returns an object of the `BodyGenerator` class, and trying to access the strings gives the error `warning: [static] static variable should be qualified by type name, BodyGenerator, instead of by an expression`. So should I import the entire `BodyGenerator` class into the tests seperately so as to access the strings via `BodyGenerator.textFileContents`? Or should this method be left non-static?

Comment (thc202) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:132]: Import `BodyGenerator` as needed (note that constants should be all uppercase and words separated with underscores).

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java (hunk 3 above)
Comment (pranavsaxena17) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java:72]: @thc202 Should I address this in `RequestModelConverter` then?

Comment (thc202) [reply addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java:72]: Yes.

**Comment by pranavsaxena17** [comment]
@thc202 Should I remove the endpoint from all definitions?(Even the ones which I did not change)

**Comment by thc202** [comment]
I'm +1 to that (worth doing in its own commit, before the other changes).

**Comment by pranavsaxena17** [comment]
@thc202 Done

**Comment by kingthorin** [comment]
@pranavsaxena17 I believe @thc202's comment was meant to imply that you should use Interactive Rebase to specifically order the changes and squash commits.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java [hunk 11]
```diff
@@ -42,6 +44,17 @@ public RequestModel convert(OperationModel operationModel, Generators generators
...
+        for (HttpHeaderField hhf : headers) {
+            if (hhf.getValue().contains("multipart/form-data")) {
+                String requestBody = requestModel.getBody();
+                String boundary = requestBody.substring(2, 38);
+                HttpHeaderField boundaryHeader =
+                        new HttpHeaderField(
+                                hhf.getName(), hhf.getValue() + "; boundary=" + boundary);
+                headers.set(headers.indexOf(hhf), boundaryHeader);
+            }
+        }
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/RequestModelConverter.java:57]: Preferable to do this when generating the headers, the body was already generated so it can make use of it.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 12]
```diff
@@ -122,6 +134,13 @@ private String generateFromArraySchema(ArraySchema schema) {
         return createJsonArrayWith(generate(schema.getItems()));
     }
 
+    private static String generateFromBinarySchema(BinarySchema schema, boolean isImage) {
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:137]: `image`

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java [hunk 13]
```diff
@@ -66,9 +66,7 @@ public Requestor(int initiator) {
                 httpRequest
                         .getRequestHeader()
                         .setContentLength(httpRequest.getRequestBody().length());
-
                 try {
-
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/network/Requestor.java:71]: Revert.

File: addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 14]
```diff
@@ -217,6 +236,107 @@ public String generateForm(Schema<?> schema) {
         return "";
     }
 
+    @SuppressWarnings("serial")
+    private static final Map<Element, String> MULTIPARTSYNTAX =
```
Comment (thc202) [inline addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:240]: I'd remove this.

File: addOns/openapi/CHANGELOG.md [hunk 15]
```diff
@@ -15,6 +18,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
```
Comment (thc202) [inline addOns/openapi/CHANGELOG.md:21]: Needs to be removed.

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]
